lib/core/src/Cardano/Wallet.hs

```diff
@@ -750,73 +752,60 @@ listUtxoStatistics ctx wid = do
 -- network tip is reached or until failure.
 restoreWallet
     :: forall ctx s k.
-        ( HasLogger WalletLog ctx
-        , HasNetworkLayer ctx
+        ( HasNetworkLayer ctx
         , HasDBLayer s k ctx
         , HasGenesisData ctx
+        , HasLogger WalletWorkerLog ctx
         , IsOurs s Address
         , IsOurs s RewardAccount
         )
     => ctx
     -> WalletId
     -> ExceptT ErrNoSuchWallet IO ()
 restoreWallet ctx wid = db & \DBLayer{..} -> do
-    cps <- liftIO $ atomically $ listCheckpoints (PrimaryKey wid)
-    let forward bs h = run $ do
-            restoreBlocks @ctx @s @k ctx wid bs h
-    liftIO (follow nw tr cps forward (view #header)) >>= \case
-        FollowFailure ->
-            restoreWallet @ctx @s @k ctx wid
-        FollowRollback point -> do
-            rollbackBlocks @ctx @s @k ctx wid point
-            restoreWallet @ctx @s @k ctx wid
-        FollowDone ->
-            pure ()
-
+    let readCps = liftIO $ atomically $ listCheckpoints (PrimaryKey wid)
+    let forward bs h innerTr = run $ do
+            restoreBlocks @ctx @s @k ctx innerTr wid bs h
+    let backward = runExceptT . rollbackBlocks @ctx @s @k ctx wid
+    liftIO $ follow nw tr readCps forward backward RetryOnExceptions st (view #header)
   where
+    (_, _, st) = ctx ^. genesisData
     db = ctx ^. dbLayer @s @k
     nw = ctx ^. networkLayer
-    tr = contramap MsgFollow (ctx ^. logger @WalletLog)
+    tr = contramap MsgFollow $ ctx ^. logger
 
     run :: ExceptT ErrNoSuchWallet IO () -> IO (FollowAction ErrNoSuchWallet)
     run = fmap (either ExitWith (const Continue)) . runExceptT
 
 -- | Rewind the UTxO snapshots, transaction history and other information to a
 -- the earliest point in the past that is before or is the point of rollback.
 rollbackBlocks
-    :: forall ctx s k.
-        ( HasLogger WalletLog ctx
-        , HasDBLayer s k ctx
-        )
+    :: forall ctx s k. (HasDBLayer s k ctx)
     => ctx
     -> WalletId
     -> SlotNo
-    -> ExceptT ErrNoSuchWallet IO ()
+    -> ExceptT ErrNoSuchWallet IO SlotNo
 rollbackBlocks ctx wid point = db & \DBLayer{..} -> do
-    lift $ traceWith tr $ MsgTryingRollback point
-    point' <- mapExceptT atomically $ rollbackTo (PrimaryKey wid) point
-    lift $ traceWith tr $ MsgRolledBack point'
+    mapExceptT atomically $ rollbackTo (PrimaryKey wid) point
   where
     db = ctx ^. dbLayer @s @k
-    tr = ctx ^. logger @WalletLog
 
 -- | Apply the given blocks to the wallet and update the wallet state,
 -- transaction history and corresponding metadata.
 restoreBlocks
     :: forall ctx s k.
-        ( HasLogger WalletLog ctx
-        , HasDBLayer s k ctx
+        ( HasDBLayer s k ctx
         , HasNetworkLayer ctx
-        , HasGenesisData ctx
         , IsOurs s Address
         , IsOurs s RewardAccount
         )
     => ctx
+    -> Tracer IO WalletFollowLog
```

Comment:
The idea was to have `follow` provide this tracer through the callback, such that it could intercept and aggregate on this "inner" trace too.

I didn't end up doing that, but I think the concept is sound and something that could be done in the future.